File: src/programmers/internal/json_schema_alias.ts
import { OpenApi } from "@samchon/openapi";

import { MetadataAlias } from "../../schemas/metadata/MetadataAlias";

import { json_schema_description } from "./json_schema_description";
import { json_schema_object } from "./json_schema_object";
import { json_schema_station } from "./json_schema_station";
import { json_schema_title } from "./json_schema_title";

export const json_schema_alias = <BlockNever extends boolean>(props: {
  blockNever: BlockNever;
  components: OpenApi.IComponents;
  alias: MetadataAlias;
}): OpenApi.IJsonSchema.IReference[] => {
  if (
    props.alias.type.value.size() === 1 &&
    props.alias.type.value.objects.length === 1
  )
    return json_schema_object({
      components: props.components,
      object: props.alias.type.value.objects[0]!,
    }) as OpenApi.IJsonSchema.IReference[];

  const $ref: string = `#/components/schemas/${props.alias.type.name}`;
  if (props.components.schemas?.[props.alias.type.name] === undefined) {
    // TEMPORARY ASSIGNMENT
    props.components.schemas ??= {};
    props.components.schemas[props.alias.type.name] = {};

    // GENERATE SCHEMA
    const schema: OpenApi.IJsonSchema | null = json_schema_station({
      blockNever: props.blockNever,
      components: props.components,
      attribute: {
        deprecated:
          props.alias.type.jsDocTags.some((tag) => tag.name === "deprecated") ||
          undefined,
        title: json_schema_title(props.alias.type),
        description: json_schema_description(props.alias.type),
      },
      metadata: props.alias.type.value,
    });
    if (schema !== null)
      Object.assign(props.components.schemas[props.alias.type.name]!, schema);
  }
  return [{ $ref }];
};

File: packages/typia/native/core/programmers/iterate/json_schema_alias.go
package iterate

import nativemetadata "github.com/samchon/typia/packages/typia/native/core/schemas/metadata"

func json_schema_alias(props struct {
	blockNever bool
	components *OpenApi_IComponents
	alias      *nativemetadata.MetadataAlias
}) []JsonSchema {
	if props.alias.Type.Value.Size() == 1 && len(props.alias.Type.Value.Objects) == 1 {
		return json_schema_object(struct {
			components *OpenApi_IComponents
			object     *nativemetadata.MetadataObject
		}{
			components: props.components,
			object:     props.alias.Type.Value.Objects[0],
		})
	}

	ref := "#/components/schemas/" + props.alias.Type.Name
	if props.components.Schemas == nil {
		props.components.Schemas = map[string]JsonSchema{}
	}
	if _, ok := props.components.Schemas[props.alias.Type.Name]; ok == false {
		props.components.Schemas[props.alias.Type.Name] = JsonSchema{}

		attribute := JsonSchema{}
		if json_schema_has_tag(props.alias.Type.JsDocTags, "deprecated") {
			attribute["deprecated"] = true
		}
		if title := json_schema_title(struct {
			description *string
			jsDocTags   []nativemetadata.IJsDocTagInfo
		}{
			description: props.alias.Type.Description,
			jsDocTags:   props.alias.Type.JsDocTags,
		}); title != nil {
			attribute["title"] = *title
		}
		if description := json_schema_description(struct {
			description *string
			jsDocTags   []nativemetadata.IJsDocTagInfo
		}{
			description: props.alias.Type.Description,
			jsDocTags:   props.alias.Type.JsDocTags,
		}); description != nil {
			attribute["description"] = *description
		}
		schema := json_schema_station(json_schema_station_props{
			blockNever: props.blockNever,
			components: props.components,
			attribute:  attribute,
			metadata:   props.alias.Type.Value,
		})
		if schema != nil {
			json_schema_jsDocTags(schema, props.alias.Type.JsDocTags)
			json_schema_assign(props.components.Schemas[props.alias.Type.Name], schema)
		}
	}
	return []JsonSchema{{"$ref": ref}}
}
